File: backend/core/timezone_utils.py
"""
Timezone utility functions for handling UTC conversion
"""
from datetime import datetime, time, timezone
from typing import Optional
import pytz


def convert_local_time_to_utc(local_time_str: str, user_timezone: str = 'UTC') -> str:
    """
    Convert a local time string (HH:MM format) to UTC time string.
    
    Args:
        local_time_str: Time in HH:MM format (e.g., "09:00")
        user_timezone: User's timezone (e.g., 'America/New_York', 'Europe/London')
    
    Returns:
        UTC time string in HH:MM format
    """
    try:
        print(f"DEBUG: Converting {local_time_str} from {user_timezone} to UTC")
        
        # Parse the time string
        hour, minute = map(int, local_time_str.split(':'))
        local_time = time(hour, minute)
        
        # Get the user's timezone
        user_tz = pytz.timezone(user_timezone)
        
        # Create a datetime object for today with the local time
        today = datetime.now(user_tz).date()
        local_datetime = user_tz.localize(datetime.combine(today, local_time))
        
        # Convert to UTC
        utc_datetime = local_datetime.astimezone(pytz.UTC)
        
        result = utc_datetime.strftime('%H:%M')
        print(f"DEBUG: Converted {local_time_str} ({user_timezone}) to {result} UTC")
        
        # Return as HH:MM string
        return result
    except Exception as e:
        print(f"DEBUG: Conversion failed: {e}")
        # If conversion fails, assume it's already UTC
        return local_time_str


def convert_utc_time_to_local(utc_time_str: str, user_timezone: str = 'UTC') -> str:
    """
    Convert a UTC time string (HH:MM format) to local time string.
    
    Args:
        utc_time_str: UTC time in HH:MM format (e.g., "14:00")
        user_timezone: User's timezone (e.g., 'America/New_York', 'Europe/London')
    
    Returns:
        Local time string in HH:MM format
    """
    try:
        # Parse the UTC time string
        hour, minute = map(int, utc_time_str.split(':'))
        utc_time = time(hour, minute)
        
        # Get UTC timezone
        utc_tz = pytz.UTC
        
        # Create a datetime object for today with the UTC time
        today = datetime.now(utc_tz).date()
        utc_datetime = utc_tz.localize(datetime.combine(today, utc_time))
        
        # Convert to user's timezone
        user_tz = pytz.timezone(user_timezone)
        local_datetime = utc_datetime.astimezone(user_tz)
        
        # Return as HH:MM string
        return local_datetime.strftime('%H:%M')
    except Exception as e:
        # If conversion fails, return as is
        return utc_time_str


def get_user_timezone_from_request(request_headers: dict) -> str:
    """
    Extract user timezone from request headers or return default.
    
    Args:
        request_headers: Request headers dictionary
    
    Returns:
        User timezone string (defaults to 'UTC')
    """
    # Try to get timezone from headers
    timezone_header = request_headers.get('x-user-timezone') or request_headers.get('timezone')
    if timezone_header:
        try:
            # Validate timezone
            pytz.timezone(timezone_header)
            return timezone_header
        except pytz.exceptions.UnknownTimeZoneError:
            pass
    
    # Default to UTC if no valid timezone found
    return 'UTC'


def convert_time_blocks_to_utc(time_blocks: list, user_timezone: str = 'UTC') -> list:
    """
    Convert a list of time blocks from local time to UTC.
    
    Args:
        time_blocks: List of time block dictionaries with 'start' and 'end' keys
        user_timezone: User's timezone
    
    Returns:
        List of time blocks with UTC times
    """
    utc_blocks = []
    for block in time_blocks:
        utc_blocks.append({
            'start': convert_local_time_to_utc(block['start'], user_timezone),
            'end': convert_local_time_to_utc(block['end'], user_timezone)
        })
    return utc_blocks


def convert_time_slots_to_utc(time_slots: list, user_timezone: str = 'UTC') -> list:
    """
    Convert a list of time slots from local time to UTC.
    
    Args:
        time_slots: List of time slot dictionaries with 'time' and 'enabled' keys
        user_timezone: User's timezone
    
    Returns:
        List of time slots with UTC times
    """
    utc_slots = []
    for slot in time_slots:
        utc_slots.append({
            'time': convert_local_time_to_utc(slot['time'], user_timezone),
            'enabled': slot['enabled']
        })
    return utc_slots


def convert_time_blocks_from_utc(time_blocks: list, user_timezone: str = 'UTC') -> list:
    """
    Convert a list of time blocks from UTC to local time.
    
    Args:
        time_blocks: List of time block dictionaries with 'start' and 'end' keys
        user_timezone: User's timezone
    
    Returns:
        List of time blocks with local times
    """
    local_blocks = []
    for block in time_blocks:
        local_blocks.append({
            'start': convert_utc_time_to_local(block['start'], user_timezone),
            'end': convert_utc_time_to_local(block['end'], user_timezone)
        })
    return local_blocks


def convert_time_slots_from_utc(time_slots: list, user_timezone: str = 'UTC') -> list:
    """
    Convert a list of time slots from UTC to local time.
    
    Args:
        time_slots: List of time slot dictionaries with 'time' and 'enabled' keys
        user_timezone: User's timezone
    
    Returns:
        List of time slots with local times
    """
    local_slots = []
    for slot in time_slots:
        local_slots.append({
            'time': convert_utc_time_to_local(slot['time'], user_timezone),
            'enabled': slot['enabled']
        })
    return local_slots

File: frontend/src/utils/timezoneUtils.ts
/**
 * Timezone utility functions for handling UTC conversion
 */

/**
 * Convert a local time string (HH:MM format) to UTC time string
 * @param localTimeStr - Time in HH:MM format (e.g., "09:00")
 * @param userTimezone - User's timezone (e.g., 'America/New_York', 'Europe/London')
 * @returns UTC time string in HH:MM format
 */
export function convertLocalTimeToUTC(localTimeStr: string, userTimezone: string = 'UTC'): string {
  try {
    // Parse the time string
    const [hour, minute] = localTimeStr.split(':').map(Number);
    
    // Create a date object for today
    const today = new Date();
    const year = today.getFullYear();
    const month = today.getMonth();
    const day = today.getDate();
    
    // Create a date string in ISO format
    const localDateStr = `${year}-${String(month + 1).padStart(2, '0')}-${String(day).padStart(2, '0')}T${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}:00`;
    
    // Create a date object representing the local time
    const localDate = new Date(localDateStr);
    
    // Get the timezone offset for the user's timezone
    const utcDate = new Date(localDateStr + 'Z'); // This is UTC
    const userTzDate = new Date(utcDate.toLocaleString('en-US', { timeZone: userTimezone }));
    
    // Calculate the offset (difference between UTC and user timezone)
    const offset = userTzDate.getTime() - utcDate.getTime();
    
    // Apply the offset to convert from user timezone to UTC
    const utcResult = new Date(localDate.getTime() - offset);
    
    // Return as HH:MM string
    return utcResult.toISOString().substr(11, 5);
  } catch {
    // If conversion fails, assume it's already UTC
    return localTimeStr;
  }
}

/**
 * Convert a UTC time string (HH:MM format) to local time string
 * @param utcTimeStr - UTC time in HH:MM format (e.g., "14:00")
 * @param userTimezone - User's timezone (e.g., 'America/New_York', 'Europe/London')
 * @returns Local time string in HH:MM format
 */
export function convertUTCToLocalTime(utcTimeStr: string, userTimezone: string = 'UTC'): string {
  try {
    // Parse the UTC time string
    const [hour, minute] = utcTimeStr.split(':').map(Number);
    
    // Create a date object for today with the UTC time
    const today = new Date();
    const year = today.getFullYear();
    const month = today.getMonth();
    const day = today.getDate();
    
    // Create a UTC date string
    const utcDateStr = `${year}-${String(month + 1).padStart(2, '0')}-${String(day).padStart(2, '0')}T${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}:00Z`;
    
    // Create a date object representing the UTC time
    const utcDate = new Date(utcDateStr);
    
    // Convert to the user's timezone
    const localTime = new Date(utcDate.toLocaleString('en-US', { timeZone: userTimezone }));
    
    // Return as HH:MM string
    return localTime.toTimeString().substr(0, 5);
  } catch {
    // If conversion fails, return as is
    return utcTimeStr;
  }
}

/**
 * Get user's timezone from browser
 * @returns User's timezone string
 */
export function getUserTimezone(): string {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone;
  } catch {
    return 'UTC';
  }
}

/**
 * Convert time blocks from local time to UTC
 * @param timeBlocks - Array of time block objects with 'start' and 'end' keys
 * @param userTimezone - User's timezone
 * @returns Array of time blocks with UTC times
 */
export function convertTimeBlocksToUTC(timeBlocks: Array<{start: string; end: string}>, userTimezone: string = 'UTC'): Array<{start: string; end: string}> {
  return timeBlocks.map(block => ({
    start: convertLocalTimeToUTC(block.start, userTimezone),
    end: convertLocalTimeToUTC(block.end, userTimezone)
  }));
}

/**
 * Convert time slots from local time to UTC
 * @param timeSlots - Array of time slot objects with 'time' and 'enabled' keys
 * @param userTimezone - User's timezone
 * @returns Array of time slots with UTC times
 */
export function convertTimeSlotsToUTC(timeSlots: Array<{time: string; enabled: boolean}>, userTimezone: string = 'UTC'): Array<{time: string; enabled: boolean}> {
  return timeSlots.map(slot => ({
    time: convertLocalTimeToUTC(slot.time, userTimezone),
    enabled: slot.enabled
  }));
}

/**
 * Convert time blocks from UTC to local time
 * @param timeBlocks - Array of time block objects with 'start' and 'end' keys
 * @param userTimezone - User's timezone
 * @returns Array of time blocks with local times
 */
export function convertTimeBlocksFromUTC(timeBlocks: Array<{start: string; end: string}>, userTimezone: string = 'UTC'): Array<{start: string; end: string}> {
  return timeBlocks.map(block => ({
    start: convertUTCToLocalTime(block.start, userTimezone),
    end: convertUTCToLocalTime(block.end, userTimezone)
  }));
}

/**
 * Convert time slots from UTC to local time
 * @param timeSlots - Array of time slot objects with 'time' and 'enabled' keys
 * @param userTimezone - User's timezone
 * @returns Array of time slots with local times
 */
export function convertTimeSlotsFromUTC(timeSlots: Array<{time: string; enabled: boolean}>, userTimezone: string = 'UTC'): Array<{time: string; enabled: boolean}> {
  return timeSlots.map(slot => ({
    time: convertUTCToLocalTime(slot.time, userTimezone),
    enabled: slot.enabled
  }));
}

/**
 * Format time for display with timezone info
 * @param timeStr - Time in HH:MM format
 * @param userTimezone - User's timezone
 * @returns Formatted time string with timezone
 */
export function formatTimeWithTimezone(timeStr: string, userTimezone: string = 'UTC'): string {
  try {
    const [hour, minute] = timeStr.split(':').map(Number);
    const today = new Date();
    const date = new Date(today.getFullYear(), today.getMonth(), today.getDate(), hour, minute);
    
    return date.toLocaleTimeString('en-US', {
      hour: 'numeric',
      minute: '2-digit',
      hour12: true,
      timeZone: userTimezone
    });
  } catch {
    return timeStr;
  }
}
